File: ml_service/feature_engineering.py
from __future__ import annotations

from typing import Any, Dict


PASSWORD_REUSE_RISK = {"never": 0, "sometimes": 2, "often": 5, "always": 8}
PASSWORD_STRENGTH_SCORE = {"weak": 25, "medium": 55, "strong": 80, "manager": 95}
TWO_FACTOR_SCORE = {"none": 0, "some": 40, "most": 75, "all": 100}
PROFILE_VISIBILITY_RISK = {"private": 10, "friends": 45, "public": 85}
SHARING_FREQUENCY_RISK = {"rarely": 10, "occasionally": 35, "often": 65, "very-often": 90}
COOKIE_RISK = {"reject": 10, "sometimes": 45, "accept": 80}
UNKNOWN_SITE_RISK = {"rare": 10, "occasional": 45, "frequent": 85}


def clamp(value: float, minimum: float = 0, maximum: float = 100) -> float:
    return max(minimum, min(maximum, value))


def build_privacy_features(form_data: Dict[str, Any], feature_order: list[str]) -> Dict[str, float]:
    password_reuse = form_data.get("passwordReuse", "sometimes")
    password_strength = form_data.get("passwordStrength", "medium")
    two_factor = form_data.get("twoFactorAuth", "some")
    profile_visibility = form_data.get("profileVisibility", "friends")
    sharing_frequency = form_data.get("sharingFrequency", "occasionally")
    cookie_consent = form_data.get("cookieConsent", "sometimes")
    unknown_site_frequency = form_data.get("unknownSiteFrequency", "occasional")
    unknown_downloads = form_data.get("unknownDownloads", "sometimes")
    ad_tracking = form_data.get("adTracking", "limited")

    online_accounts = float(form_data.get("onlineAccounts", 0) or 0)
    inactive_accounts = float(form_data.get("inactiveAccounts", 0) or 0)
    connected_apps = float(form_data.get("connectedApps", 0) or 0)
    shared_data_types = form_data.get("sharedDataTypes", []) or []
    login_methods = form_data.get("loginMethods", []) or []
    additional_notes = form_data.get("additionalNotes", "") or ""

    risky_domains_visited = UNKNOWN_SITE_RISK.get(unknown_site_frequency, 45) / 10
    risky_domains_visited += 4 if unknown_downloads == "often" else 2 if unknown_downloads == "sometimes" else 0

    protection_score = clamp(
        TWO_FACTOR_SCORE.get(two_factor, 40) * 0.45
        + PASSWORD_STRENGTH_SCORE.get(password_strength, 55) * 0.35
        + (20 if ad_tracking == "disabled" else 10 if ad_tracking == "limited" else 0)
    )

    exposure_index = clamp(
        online_accounts * 1.2
        + inactive_accounts * 2.5
        + connected_apps * 1.8
        + len(shared_data_types) * 4
        + PROFILE_VISIBILITY_RISK.get(profile_visibility, 45) * 0.35
    )

    risk_burden = clamp(
        exposure_index * 0.45
        + PASSWORD_REUSE_RISK.get(password_reuse, 2) * 6
        + SHARING_FREQUENCY_RISK.get(sharing_frequency, 35) * 0.25
        + COOKIE_RISK.get(cookie_consent, 45) * 0.15
    )

    privacy_awareness = (
        2
        if len(additional_notes) > 80 or two_factor == "all"
        else 1
        if two_factor == "most" or password_strength == "manager"
        else 0
    )

    features = {
        "privacy_awareness": privacy_awareness,
        "password_reuse_count": PASSWORD_REUSE_RISK.get(password_reuse, 2),
        "risky_domains_visited": risky_domains_visited,
        "third_party_apps": connected_apps,
        "inactive_accounts": inactive_accounts,
        "breached_accounts": 0,
        "tracker_acceptance_rate": COOKIE_RISK.get(cookie_consent, 45) / 100,
        "avg_password_strength": PASSWORD_STRENGTH_SCORE.get(password_strength, 55),
        "public_profile_score": PROFILE_VISIBILITY_RISK.get(profile_visibility, 45),
        "pii_shared_frequency": SHARING_FREQUENCY_RISK.get(sharing_frequency, 35),
        "vpn_usage": 1 if any("vpn" in str(method).lower() for method in login_methods) else 0,
        "mfa_enabled": TWO_FACTOR_SCORE.get(two_factor, 40) / 100,
        "browser_security_score": clamp(100 - UNKNOWN_SITE_RISK.get(unknown_site_frequency, 45)),
        "account_age_days": 365,
        "risk_burden": risk_burden,
        "protection_score": protection_score,
        "exposure_index": exposure_index,
        "awareness_adjusted_risk": clamp(risk_burden - privacy_awareness * 8),
    }

    return {feature: float(features.get(feature, 0)) for feature in feature_order}

File: lib/ml/feature-engineering.ts
import "server-only"

import type { PrivacyFormData } from "@/lib/types"

export type PrivacyFeatureVector = Record<string, number>

const passwordReuseRisk: Record<PrivacyFormData["passwordReuse"], number> = {
  never: 0,
  sometimes: 2,
  often: 5,
  always: 8,
}

const passwordStrengthScore: Record<PrivacyFormData["passwordStrength"], number> = {
  weak: 25,
  medium: 55,
  strong: 80,
  manager: 95,
}

const twoFactorScore: Record<PrivacyFormData["twoFactorAuth"], number> = {
  none: 0,
  some: 40,
  most: 75,
  all: 100,
}

const profileVisibilityRisk: Record<PrivacyFormData["profileVisibility"], number> = {
  private: 10,
  friends: 45,
  public: 85,
}

const sharingFrequencyRisk: Record<PrivacyFormData["sharingFrequency"], number> = {
  rarely: 10,
  occasionally: 35,
  often: 65,
  "very-often": 90,
}

const cookieRisk: Record<PrivacyFormData["cookieConsent"], number> = {
  reject: 10,
  sometimes: 45,
  accept: 80,
}

const unknownSiteRisk: Record<PrivacyFormData["unknownSiteFrequency"], number> = {
  rare: 10,
  occasional: 45,
  frequent: 85,
}

const appDomainMap: Record<string, string> = {
  instagram: "instagram.com",
  facebook: "facebook.com",
  whatsapp: "whatsapp.com",
  google: "google.com",
  gmail: "gmail.com",
  amazon: "amazon.com",
  linkedin: "linkedin.com",
  netflix: "netflix.com",
  spotify: "spotify.com",
  dropbox: "dropbox.com",
  notion: "notion.so",
  canva: "canva.com",
  slack: "slack.com",
  trello: "trello.com",
  zapier: "zapier.com",
  ifttt: "ifttt.com",
}

const categoryDomainMap: Record<string, string[]> = {
  "Social Media": ["x.com", "instagram.com", "linkedin.com"],
  "Email Services": ["gmail.com", "outlook.com"],
  "Financial Apps": ["paypal.com"],
  "Shopping Platforms": ["amazon.com", "flipkart.com"],
  "Cloud Storage": ["drive.google.com", "dropbox.com"],
  "Messaging Apps": ["whatsapp.com", "telegram.org"],
  "Gaming Platforms": ["steampowered.com"],
  "Government Services": ["uidai.gov.in"],
}

function boolScore(value: boolean) {
  return value ? 1 : 0
}

function clamp(value: number, min = 0, max = 100) {
  return Math.max(min, Math.min(max, value))
}

export function extractDomainsFromFormData(data: PrivacyFormData) {
  const rawValues = [
    data.frequentSites,
    ...data.specificApps,
    ...data.thirdPartyAppNames,
  ]

  const explicitDomains = rawValues
    .flatMap((value) => value.split(/[\s,;]+/))
    .map((value) => value.trim().toLowerCase())
    .map((value) => value.replace(/^https?:\/\//, "").split("/")[0])
    .filter((value) => value.includes(".") && value.length > 3)

  const mappedAppDomains = [...data.specificApps, ...data.thirdPartyAppNames]
    .map((value) => value.trim().toLowerCase())
    .map((value) => appDomainMap[value])
    .filter(Boolean)

  const categoryDomains = data.platformCategories.flatMap((category) => categoryDomainMap[category] || [])
  const domains = [...explicitDomains, ...mappedAppDomains, ...categoryDomains]

  return Array.from(new Set(domains)).slice(0, 25)
}

export function buildPrivacyFeatureVector(data: PrivacyFormData): PrivacyFeatureVector {
  const riskyDomainsVisited =
    unknownSiteRisk[data.unknownSiteFrequency] / 10 +
    (data.unknownDownloads === "often" ? 4 : data.unknownDownloads === "sometimes" ? 2 : 0)

  const protectionScore = clamp(
    twoFactorScore[data.twoFactorAuth] * 0.45 +
      passwordStrengthScore[data.passwordStrength] * 0.35 +
      (data.adTracking === "disabled" ? 20 : data.adTracking === "limited" ? 10 : 0)
  )

  const exposureIndex = clamp(
    data.onlineAccounts * 1.2 +
      data.inactiveAccounts * 2.5 +
      data.connectedApps * 1.8 +
      data.sharedDataTypes.length * 4 +
      profileVisibilityRisk[data.profileVisibility] * 0.35
  )

  const riskBurden = clamp(
    exposureIndex * 0.45 +
      passwordReuseRisk[data.passwordReuse] * 6 +
      sharingFrequencyRisk[data.sharingFrequency] * 0.25 +
      cookieRisk[data.cookieConsent] * 0.15
  )

  const privacyAwareness =
    data.additionalNotes.length > 80 || data.twoFactorAuth === "all"
      ? 2
      : data.twoFactorAuth === "most" || data.passwordStrength === "manager"
        ? 1
        : 0

  return {
    privacy_awareness: privacyAwareness,
    password_reuse_count: passwordReuseRisk[data.passwordReuse],
    risky_domains_visited: riskyDomainsVisited,
    third_party_apps: data.connectedApps,
    inactive_accounts: data.inactiveAccounts,
    breached_accounts: 0,
    tracker_acceptance_rate: cookieRisk[data.cookieConsent] / 100,
    avg_password_strength: passwordStrengthScore[data.passwordStrength],
    public_profile_score: profileVisibilityRisk[data.profileVisibility],
    pii_shared_frequency: sharingFrequencyRisk[data.sharingFrequency],
    vpn_usage: boolScore(data.loginMethods.some((method) => method.toLowerCase().includes("vpn"))),
    mfa_enabled: twoFactorScore[data.twoFactorAuth] / 100,
    browser_security_score: clamp(100 - unknownSiteRisk[data.unknownSiteFrequency]),
    account_age_days: 365,
    risk_burden: riskBurden,
    protection_score: protectionScore,
    exposure_index: exposureIndex,
    awareness_adjusted_risk: clamp(riskBurden - privacyAwareness * 8),
  }
}
